### PR TITLE
Likert Component

### PR DESCRIPTION
**Pull request by lasseklovstad**
## Description
Hey! I Have a new component I hope can be included for our use in Elevundersøkelsen at UDIR. This component fits our needs and should be functional for others to use. It requires a list in the datamodel and is linked in the new dataModelBinding type for the likert.

PR to support pdf generation: https://github.com/Altinn/altinn-studio/pull/8441
PR to support in schema: https://github.com/Altinn/altinn-cdn/pull/95
PR to documentation: https://github.com/Altinn/altinn-studio-docs/pull/506

### Likert component desktop
You should be able to use keyboard navigation in the table. The table uses auto-layout provided by the browser. This layout is tested with windows screen reader. 
Normal state:
ny-likert-med-fokus
Error state:
ny-likert-medfeil

### Likert component mobile
Used mediaQuery: max-width:992px. In mobile view the component is rendered as normal radiobuttons.
Normal state:
ny-likert-mobil
Error state:
ny-likert-mobil-feil

### Layout.json
Layout file in your alltinn studio app:
```
{
        "id": "likert-ny",
        "type": "Group",
        "maxCount": 99,
        "children": [
          "likert-answer1"
        ],
        "textResourceBindings": {
          "title": "layoutSchoolWork",
          "description":  "På de neste spørsmålene skal du svare på i kontekst av klasserommet."
        },
        "dataModelBindings": {
          "group": "Questions"
        },
        "edit": {
          "mode": "likert",
          "filter": [
            {
              "key": "start",
              "value": 1
            },
            {
              "key": "stop",
              "value": 4
            }
          ]
        }
      },
      {
        "id": "likert-answer1",
        "type": "RadioButtons",
        "textResourceBindings": {
          "title": "likert-questions"
        },
        "dataModelBindings": {
          "simpleBinding": "Questions.Answer"
        },
        "optionsId": "frequency",
        "required": false,
        "readOnly": false
      }
```
- **edit.filter.start**: (optional) Used to slice the array of questions (inclusive), if not provided we start from index 0
- **edit.filter.end**:  (optional) Used to slice the array of questions (exclusive), if not provided we end with length of likert list 

The textresourceBings.title of the RadioButton component looks like this in the resource file:
```
{
      "id": "likert-questions",
      "value": "{0}",
      "variables": [
        {
          "key": "Questions[{0}].Id",
          "dataSource": "dataModel.default"
        }
      ]
    }
```
This will bind each question to the value in the datamodel for the current index. The value in the datamodel can support a reference back to a translation file again.

## Verification
- [x] All tests run except one snapshot test with repeating group localy. This one also failes on main branch.
- [x] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [x] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) Dette skal jeg begynne på.
- [ ] Changelog is updated with a separate linked PR
  - [ ] [altinn-app-frontend](https://docs.altinn.studio/community/changelog/app-frontend/)- Har ikke gjort dette.
  - [ ] [nuget-packages](https://docs.altinn.studio/community/changelog/app-nuget/) - Ikke relevant?

### CODE REVIEW

**Comment by ivarne** [comment]
A few extra TODOs:

- [ ] Update https://github.com/Altinn/altinn-cdn/blob/master/schemas/json/layout/layout.schema.v1.json
- [ ] Support for configuration in Altinn Studio.

**Comment by lasseklovstad** [comment]
> A few extra TODOs:
> 
> * [ ]  Update https://github.com/Altinn/altinn-cdn/blob/master/schemas/json/layout/layout.schema.v1.json
> * [ ]  Support for configuration in Altinn Studio.

https://github.com/Altinn/altinn-cdn/pull/95

**Comment by lasseklovstad** [comment]
> A few extra TODOs:
> 
> * [ ]  Update https://github.com/Altinn/altinn-cdn/blob/master/schemas/json/layout/layout.schema.v1.json
> * [ ]  Support for configuration in Altinn Studio.

Is this something you guys can help out with? Support for configuration in Altinn Studio

**Comment by nkylstad** [comment]
Hi @lasseklovstad!
Thanks for this contribution! We have added reviewing this PR to our current sprint (I saw that @lorang92 mentioned this in a Slack thread earlier today), and will do our best to get back to you as soon as possible. The changes in this PR address some new concepts that we have not supported before - as such we will need some time to review it and determine how to best get this component into our code base 😄 
We will get in touch if we need some more info or input from you!

Regarding support for configuration in Studio, I see that we are lacking documentation for this. Might want to hold off on that while we review this PR, in case there are any changes here that would affect that implementation. I have created an issue to update the docs with instructions, will try to get it prioritized. Otherwise you can message us on Slack when you're ready to start and we can guide you through it.

**Comment by nkylstad** [comment]
@lasseklovstad What does the data model set up look like in the xsd/JSONSchema data model? Do you have an example app in f.ex. Altinn Studio that I can look at?

**Comment by lasseklovstad** [comment]
> @lasseklovstad What does the data model set up look like in the xsd/JSONSchema data model? Do you have an example app in f.ex. Altinn Studio that I can look at?

@nkylstad I have an example here: https://altinn.studio/repos/udir/elevundersokelsen/src/branch/REU-466-Likert-oppsett 

The questions is populated into the model in the instantiationhandler: https://altinn.studio/repos/udir/elevundersokelsen/src/branch/REU-466-Likert-oppsett/App/logic/InstantiationHandler.cs

and the Likert json config is seen here: https://altinn.studio/repos/udir/elevundersokelsen/src/branch/REU-466-Likert-oppsett/App/ui/student-survey/layouts/school-work.json

if you want to run this app localy you need to add custom claims to the altin platform described in this PR: https://github.com/Altinn/altinn-studio/pull/8434

**Comment by lasseklovstad** [comment]
> @lasseklovstad What does the data model set up look like in the xsd/JSONSchema data model? Do you have an example app in f.ex. Altinn Studio that I can look at?

@nkylstad I pushed a change in the altinn so that you can run it without the claims locally for testing. https://altinn.studio/repos/udir/elevundersokelsen/src/branch/REU-466-Likert-oppsett

File: src/altinn-app-frontend/package.json
Comment (haakemon) [inline src/altinn-app-frontend/package.json:31]: Do we need this dependency? We have other tests that set screen size to simulate mobile screens, f.ex https://github.com/Altinn/app-frontend-react/blob/main/src/altinn-app-frontend/src/components/base/DatepickerComponent.test.tsx#L46-L61
This could be extracted to a generic testutil f.ex (not saying you need to do this as part of this PR :) ) but I would like to avoid adding a dependency that hasnt seen updates in 8 years :)

**Review by nkylstad — COMMENTED** [review]
Thanks again for this contribution! Here is our initial feedback, we can discuss further on slack 😄 

Overall, we would like this component to be implemented as a specialized view of our repeating group concept. Read more here: https://docs.altinn.studio/app/development/ux/fields/grouping/repeating/
This is because we have already handled things like indexes and validation on lists of questions within this concept. 

Ideally, this component could be solved with a config like this:
```json
{
  "id": "likert",
  "type":  "Group",
  "maxCount": 99,
  "children": [
    "likert-question",
    "likert-answer"
  ],
  "textResourceBindings": {
    "title": "layoutSchoolWork"
  },
  "dataModelBindings": {
    "group": "Questions"
  },
  "edit": {
    "mode": "likert", // NEW, this is how we decide to use likert view instead of standard
    "filter": [ // would need custom implementation of this in likert view
      {"key": "start", "value": 1},
      {"key": "stop", "value": 4}
    ]
  }
},
{
  "id": "likert-answer",
  "type": "RadioButtons",
  "textResourceBindings": {
    "title": "textFromDataModel" // use variables in text to populate
  },
  "dataModelBindings": {
    "simpleBinding": "Questions.Answer"
  },
  "optionsId": "frequency",
  "required": false,
  "readOnly": false,
}
```

This is re-using our existing config. Things like indexes, saving data to the correct place in the list, validation, etc. is handled out of the box (assuming we can re-use some stuff from existing repeating group implementation).
Another reason to solve it like this is that it also supports replacing the radio buttons with another scale-type component (if we have any in the future) without having to make a lot of changes.

We can generate this form layout setup as a Likert [widget](https://docs.altinn.studio/app/development/ux/widgets/) in Altinn Studio, to make setting it up in a form easier.

Repeating groups are set up in `src/features/form/containers/GroupContainer.tsx`, we are thinking that a new component under this container is probably the way to go.

Have a look at the linked docs, and the feedback above, and let's talk more on Slack about the best way forward!

File: src/shared/src/components/molecules/AltinnTableHeader.tsx
Comment (mjulstein) [reply src/shared/src/components/molecules/AltinnTableHeader.tsx:20]: No it's not `'0 0 4px 36px'´ would be.

File: src/altinn-app-frontend/src/components/base/RadioButtons/radioButtonsUtils.ts
Comment (nkylstad) [inline src/altinn-app-frontend/src/components/base/RadioButtons/radioButtonsUtils.ts:67]: See warning below

File: src/altinn-app-frontend/src/features/form/components/Label.tsx
Comment (nkylstad) [inline src/altinn-app-frontend/src/features/form/components/Label.tsx:44]: Is there a need for this change? I find that having the checks here in Label.tsx gives a much more verbose and clear picture of when the different indicators are shown. We also let the indicator components fetch from the store directly to avoid unnecessary prop drilling. But I'm open to making this change if there are good reasons for it 😄

File: src/shared/src/components/organisms/AltinnTable.tsx
Comment (nkylstad) [inline src/shared/src/components/organisms/AltinnTable.tsx:34]: See warnings below

**Review by nkylstad — COMMENTED** [review]
Starting to look really good! I have commented requesting one little change before I'm totally happy with this 😇 I will trigger automated test pipelines etc. now to uncover any other small things that might need fixing before we merge and test.

File: src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx
Comment (lasseklovstad) [reply src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx]: Desktop/mobile view is decided in The Container component (RepeatingGroupsLikertContainer.tsx). It could be decided in the Likert component, but it would have to be in both places. I did it in the Container component because the desktop view requires a table layout, where the likert component is a table row. The row is dependent on that there exists a table layout containing it. In my opinion this is best to control on the container.

File: src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx
Comment (lasseklovstad) [reply src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx]: good question. Do actually know the difference between the two props? Are they the same? I can't see validationMessages being passed down as a prop anywhere.

File: src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx
Comment (lasseklovstad) [reply src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx]: The reason for this is to reference the headercolumn value of the current option of the radiobutton in LikertComponent.tsx line52 and 69:
```
// column label must reference correct id of header in table
const colLabelId = `${groupContainerId}-likert-columnheader-${colIndex}`;

<StyledRadio
                inputProps={{
                  'aria-labelledby': `${rowLabelId} ${colLabelId}`,
                  id: inputId,
                  role: 'radio',
                  name: rowLabelId,
                  'aria-checked': isChecked,
                }}
                checked={isChecked}
                onChange={handleChange}
                onBlur={handleBlur}
                value={option.value}
              />

```

Comment (lasseklovstad) [reply src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx]: I could use the aria-label attribute instead of aria-labeledby, but this is not recommended to do when you have a visible html element you can reference.

File: src/altinn-app-frontend/src/features/form/components/Label.tsx
Comment (lasseklovstad) [reply src/altinn-app-frontend/src/features/form/components/Label.tsx:44]: The reason I changed it was to avoid having to copy the logic for when to show hide the Optional indicator every time you want to use it. This logikk is written in 3 different components (AdressLabel, Label and Legend) and in my opinion it should be onyl written one place :P Another option could be to have a helper function for when to hid/show the OptionalIndicator outside the component.

I can of course remove it if you don't see the need for it :) You decide!

File: src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx
Comment (nkylstad) [reply src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx]: Hmm I think that some old comments I added have made their way in here, sorry about that. The comments on code that is marked as "outdated" can be ignored. See the comment below on using `layout` prop instead, and f.eks. adding a `table` value to the LayoutStyle enum. Since mobile view is basically just the same as a regular radiobuttons group, we really only need 1 setting?

File: src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx
Comment (nkylstad) [reply src/altinn-app-frontend/src/components/base/RadioButtonsContainerComponent.tsx]: Yep, I see that we need the group container id, I would keep that as is :)

File: src/shared/src/components/organisms/AltinnTable.tsx
Comment (nkylstad) [reply src/shared/src/components/organisms/AltinnTable.tsx:34]: Ah, ok, I was a bit quick here and didn't realize that's what that line was doing 😅 . Yep, then you can go ahead and update eslint config with the suggestion above :)

**Review by lorang92 — APPROVED** [review]
👏 👏 👏 

Nice!

File: src/altinn-app-frontend/src/features/form/components/Label.tsx
Comment (nkylstad) [reply src/altinn-app-frontend/src/features/form/components/Label.tsx:44]: Ok, I see your point :) Let's leave it as you have implemented it.

File: src/altinn-app-frontend/src/components/index.ts
Comment (nkylstad) [reply src/altinn-app-frontend/src/components/index.ts:169]: Hmm excellent question.. You're probably right, I think these were used in Studio to make sure that all components in the layout had the correct properties. @lasseklovstad I think you can just remove `customProperties` here.

**Review by nkylstad — APPROVED** [review]
Fantastic job! 🥇 
I will do some manual testing now to verify before merge.
There is also a merge conflict here that needs to be fixed.

**Comment by lasseklovstad** [comment]
@nkylstad 
Sounds good!

**Pull request merged**
